Removing structurally non-observable variables

In many cases, when dealing with performance-critical models, there are some parts that are not of interest for some simulations, so their computation should be skipped. One way to handle this, which is used e.g. for visualization features in the Modelica MultiBody library, is via conditional components, but this is cumbersome and needs to be programmed explicitly. 

A better idea is to declare what are the outputs of interest for the simulation, e.g. via annotations or making them top-level outputs of the model, performing a structural observability analysis, and only keep the parts of the model that are strictly necessary to compute those outputs. In the case of electrical system, one could get rid of several auxiliary variables (e.g. angles and moduli of voltage and current phasors) that are not of interest and take some time to compute because of the arctan and square root functions.

A similar problem is already solved by the backend to separate equations that are not needed to compute the derivatives and the zero-crossing functions,, which are only solved when the step is accepted. However, at that point _all_ other variables are computed.

An algorithm to perform this operation is available in [this paper](http://dx.doi.org/10.3384/ecp11063784). It can be easily implemented as an extension of existing matching algorithms in the OpenModelica backend. This could also be used in more sophisticated cases, e.g. to remove off-plane dynamic equations (and states!) from a planar model built with the 3D MultiBody library, provided that the sensors are only capturing in-plane motion and forces.

While trying to come up with a model to explain my point of structural changes to the initial system due to removed variables, i stumbled into a minimal model  that shows the problem pretty well, it seems like.

```
model test_remove
  Real x;
  output Real y,z;
initial equation
  x = y;
equation
  x = sin(time);
  der(z) = y - 3;
  y = z^2 - 10;
end test_remove;
```

My point being is that in the simulation system `x` can be removed, but from the initial system it cannot, since `y` depends on `x` in that case. We don't seem to check for that. Simulating this model works fine, but adding `-d=evalOutputOnly` returns
```
test_remove_06inz.c:20:56: error: use of undeclared identifier 'x'
  (data->localData[0]->realVars[3] /* y variable */) = x;
                                                       ^
1 error generated.
make: *** [<builtin>: test_remove_06inz.o] Error 1
make: *** Waiting for unfinished jobs....
Compilation process failed. Exited with code 2.
```
which is a direct of `x` being removed but being needed for initialization. The module is quite complicated, uses some tasking graphs which i would need to understand first. I don't know if it's easy to just also check the initial system.

--- a/OMCompiler/Compiler/BackEnd/BackendDAEOptimize.mo
+++ b/OMCompiler/Compiler/BackEnd/BackendDAEOptimize.mo
@@ -5752,7 +5752,7 @@ author: Waurich TUD 09/2015"
 protected
   Integer size, nVars, nEqs;
   array<Integer> ass1,ass2, varVisited;
-  list<Integer> outputVarIndxs, stateIndxs, stateTasks, stateTasks1 , outputTasks, predecessors, tasks, varIdcs, eqIdcs, stateDerIdcs;
+  list<Integer> outputVarIndxs, allVarIndxs, stateIndxs, stateTasks, stateTasks1 , outputTasks, predecessors, tasks, varIdcs, eqIdcs, stateDerIdcs;
   list<BackendDAE.StrongComponent> comps, compsNew, addComps;
   BackendDAE.StrongComponent comp;
   BackendDAE.EqSystem syst;
@@ -5802,6 +5802,7 @@ algorithm
     //get output variables
     BackendDAE.EQSYSTEM(orderedVars = vars) := syst;
     varLst := BackendVariable.varList(vars);
+    allVarIndxs := BackendVariable.getVarIndexFromVars(varLst,vars);
     varLst := List.filterOnTrue(varLst,BackendVariable.isVarOnTopLevelAndOutput);
 
     if not listEmpty(varLst) then
@@ -5864,10 +5865,13 @@ algorithm
         eqIndexLst := listAppend(eqIndLst,eqIndexLst);
       end for;
 
-      // causalize again
+      // find new and unneeded vars and equations
       syst.orderedVars := BackendVariable.listVar1(listReverse(varLstNew));
       syst.orderedEqs := BackendEquation.listEquation(listReverse(eqLstNew));
+      vars := BackendVariable.deleteVars(syst.orderedVars, vars);
+      eqs := BackendEquation.deleteList(eqs, eqIndexLst);
 
+      // causalize again
       syst.m := NONE();
       syst.mT := NONE();
       syst.matching := BackendDAE.NO_MATCHING();

--- a/OMCompiler/Compiler/BackEnd/BackendDAEOptimize.mo
+++ b/OMCompiler/Compiler/BackEnd/BackendDAEOptimize.mo
@@ -5752,7 +5752,7 @@ author: Waurich TUD 09/2015"
 protected
   Integer size, nVars, nEqs;
   array<Integer> ass1,ass2, varVisited;
-  list<Integer> outputVarIndxs, allVarIndxs, stateIndxs, stateTasks, stateTasks1 , outputTasks, predecessors, tasks, varIdcs, eqIdcs, stateDerIdcs;
+  list<Integer> outputVarIndxs, stateIndxs, stateTasks, stateTasks1 , outputTasks, predecessors, tasks, varIdcs, eqIdcs, stateDerIdcs;
   list<BackendDAE.StrongComponent> comps, compsNew, addComps;
   BackendDAE.StrongComponent comp;
   BackendDAE.EqSystem syst;
@@ -5802,7 +5802,6 @@ algorithm
     //get output variables
     BackendDAE.EQSYSTEM(orderedVars = vars) := syst;
     varLst := BackendVariable.varList(vars);
-    allVarIndxs := BackendVariable.getVarIndexFromVars(varLst,vars);
     varLst := List.filterOnTrue(varLst,BackendVariable.isVarOnTopLevelAndOutput);
 
     if not listEmpty(varLst) then
@@ -5865,13 +5864,10 @@ algorithm
         eqIndexLst := listAppend(eqIndLst,eqIndexLst);
       end for;
 
-      // find new and unneeded vars and equations
+      // causalize again
       syst.orderedVars := BackendVariable.listVar1(listReverse(varLstNew));
       syst.orderedEqs := BackendEquation.listEquation(listReverse(eqLstNew));
-      vars := BackendVariable.deleteVars(syst.orderedVars, vars);
-      eqs := BackendEquation.deleteList(eqs, eqIndexLst);
 
-      // causalize again
       syst.m := NONE();
       syst.mT := NONE();
       syst.matching := BackendDAE.NO_MATCHING();

--- a/OMCompiler/Compiler/BackEnd/BackendDAEOptimize.mo
+++ b/OMCompiler/Compiler/BackEnd/BackendDAEOptimize.mo
@@ -87,6 +87,7 @@ import SCode;
 import SynchronousFeatures;
 import Tearing;
 import Types;
+import UnorderedMap;
 import Util;
 import Values;
 
@@ -5777,6 +5778,8 @@ protected
   Integer systemNumber=0, numberOfSystems;
 
   list<Integer> eqIndLst, eqIndexLst = {};
+  UnorderedMap<DAE.ComponentRef, DAE.Exp> der_replacement;
+  BackendDAE.Var derVar;
 
   constant Boolean debug = false;
 algorithm
@@ -5897,8 +5900,21 @@ algorithm
       if debug then print("No output variables in this system ("+intString(systemNumber)+"/"+intString(numberOfSystems)+")\n"); end if;
     end if;
 
+    // make unneeded state derivatives and add them to unneeded vars
+    der_replacement := UnorderedMap.new<DAE.Exp>(ComponentReference.hashComponentRef, ComponentReference.crefEqual);
+    for state in BackendVariable.varList(vars) loop
+      if BackendVariable.isStateVar(state) then
+        derVar := BackendVariable.makeVar(ComponentReference.prependStringCref("$DER_REM_", state.varName));
+        UnorderedMap.add(state.varName, Expression.crefExp(derVar.varName), der_replacement);
+        vars := BackendVariable.addVar(derVar, vars);
+      end if;
+    end for;
+
+    // replace unneeded der() calls with derivative crefs
+    eqs := BackendEquation.traverseEquationArray_WithUpdate(eqs, function BackendEquation.traverseExpsOfEquation(inFunc = replaceDerCallOutputsOnly), der_replacement);
+
     // make unneeded vars parameters and equations initial equations
-    (vars, _) := BackendVariable.traverseBackendDAEVarsWithUpdate(vars, BackendVariable.makeParamFixed, false);
+    (vars, _) := BackendVariable.traverseBackendDAEVarsWithUpdate(vars, BackendVariable.makeParamOutputsOnly, false);
     (eqs, _) := BackendEquation.traverseEquationArray_WithUpdate(eqs,BackendEquation.setEquationKind, BackendDAE.INITIAL_EQUATION());
 
     // add unneeded variables and equations
@@ -5919,6 +5935,18 @@ algorithm
   b := intLt(arrayGet(varArr,idx),0);
 end stateVarIsNotVisited;
 
+protected function replaceDerCallOutputsOnly
+  input output DAE.Exp exp;
+  input output UnorderedMap<DAE.ComponentRef, DAE.Exp> der_replacement;
+algorithm
+  exp := match exp
+    local
+      DAE.ComponentRef cr;
+    case DAE.CALL(path = Absyn.IDENT("der"), expLst = {DAE.CREF(cr)})
+      then UnorderedMap.getOrDefault(cr, der_replacement, exp);
+    else exp;
+  end match;
+end replaceDerCallOutputsOnly;
 
 // =============================================================================
 // section for initOptModule >>inlineHomotopy<<

--- a/OMCompiler/Compiler/BackEnd/BackendVariable.mo
+++ b/OMCompiler/Compiler/BackEnd/BackendVariable.mo
@@ -1248,14 +1248,18 @@ algorithm
   var.varKind := BackendDAE.PARAM();
 end makeParam;
 
-public function makeParamFixed
+public function makeParamOutputsOnly
   "Change variable to parameter"
   input output BackendDAE.Var var;
   input output Boolean fixed; // also output for traversing
 algorithm
   var.varKind := BackendDAE.PARAM();
   var := setVarFixed(var, fixed);
-end makeParamFixed;
+  var.values := if isSome(var.values) then var.values else SOME(getVariableAttributefromType(var.varType));
+  if isNone(DAEUtil.getFixedAttr(var.values)) then
+    var.values := DAEUtil.setFixedAttr(var.values, SOME(DAE.BCONST(fixed)));
+  end if;
+end makeParamOutputsOnly;
 
 public function isParamOrConstant
 "Return true if variable is parameter or constant"

--- a/OMCompiler/Compiler/FrontEnd/DAEUtil.mo
+++ b/OMCompiler/Compiler/FrontEnd/DAEUtil.mo
@@ -1296,6 +1296,21 @@ algorithm
   end match;
 end setFixedAttr;
 
+public function getFixedAttr "
+  retrieves the protected attribute form VariableAttributes."
+  input Option<DAE.VariableAttributes> attr;
+  output Option<DAE.Exp> isFixed;
+algorithm
+  isFixed := match(attr)
+    case (SOME(DAE.VAR_ATTR_REAL(fixed=isFixed))) then isFixed;
+    case (SOME(DAE.VAR_ATTR_INT(fixed=isFixed))) then isFixed;
+    case (SOME(DAE.VAR_ATTR_BOOL(fixed=isFixed))) then isFixed;
+    case (SOME(DAE.VAR_ATTR_STRING(fixed=isFixed))) then isFixed;
+    case (SOME(DAE.VAR_ATTR_ENUMERATION(fixed=isFixed))) then isFixed;
+    else NONE();
+  end match;
+end getFixedAttr;
+
 public function setFinalAttr "
   sets the start attribute. If NONE(), assumes Real attributes."
   input Option<DAE.VariableAttributes> attr;

